### PR TITLE
Make CollectionProxy#proxy_association public API

### PR DESCRIPTION
**Pull request by zzak**
Fixes #51395

Screenshot from 2024-05-27 19-46-53

### CODE REVIEW

File: activerecord/lib/active_record/associations/collection_proxy.rb [hunk 1]
```diff
@@ -928,7 +928,20 @@ def include?(record)
         !!@association.include?(record)
       end
 
-      def proxy_association # :nodoc:
+      # Returns the association for the collection.
```
Comment (p8) [inline activerecord/lib/active_record/associations/collection_proxy.rb:931]: In this case association typically would mean the collection of pets?
Maybe:
```suggestion
      # Returns the association object for the collection.
```

File: activerecord/lib/active_record/associations/collection_proxy.rb [hunk 2]
```diff
@@ -928,7 +928,20 @@ def include?(record)
...
+      #
+      #   class Person < ActiveRecord::Base
+      #     has_many :pets
+      #   end
+      #
+      #   person.pets.proxy_association
+      #   # => #<ActiveRecord::Associations::HasManyAssociation owner="#<Person:0x00>">
+      #
+      # Returns the same object as <tt>person.association(:pets)</tt>,
```
Comment (p8) [inline activerecord/lib/active_record/associations/collection_proxy.rb:940]: `association` isn't a public method so I don't think we should mention it.
Maybe:
```suggestion
      # Returns the Association object,
```

**Comment by rafaelfranca** [comment]
Yes, we should

File: activerecord/lib/active_record/associations/collection_proxy.rb (hunk 2 above)
Comment (zzak) [reply activerecord/lib/active_record/associations/collection_proxy.rb:940]: It's documented here:
https://github.com/rails/rails/blob/f6918a5b5d0b8c3a700513c2624cc195e0a78d14/activerecord/lib/active_record/associations.rb#L719-L732

I linked to this section from the method docs, but that is a good point.

> Returns the Association object,
Association doesn't exist and Associations is also private. :thinking:

Comment (rafaelfranca) [reply activerecord/lib/active_record/associations/collection_proxy.rb:940]: We should make `association` public as well along with owner, reflection and target

**Comment by rafaelfranca** [comment]
I'll merge this because it is ready, but we should make `association` and the methods we document extension should use public in another PR.

**Comment by p8** [comment]
@rafaelfranca Should `#target` also be documented or should it be `:nodoc:`? https://github.com/rails/rails/blob/1e026dbf504547654eb03325448a7fffd575f923/activerecord/lib/active_record/associations/collection_proxy.rb#L40

**Comment by zzak** [comment]
That was my assumption based on https://github.com/rails/rails/pull/51925#discussion_r1617675975.

> We should make association public as well along with owner, reflection and target on that object

@p8 Do you want to open the PR?

**Pull request merged**